doc/designs/event-queue.md

```diff
@@ -0,0 +1,181 @@
+
+Events may be dynamically allocated.  In that case, ownership is passed to
+the event dispatch functionality (see below), and the event will be
+automatically destroyed when not needed any more.
+
+Events may also be embedded in another structure, as a static variable, or
+as an local variable in a function.  In that case, owndership remains with
+that structure or that variable.
+
+### Event queue
+
+An event queue is a structure holding the following:
+
+-   A reference to a method structure to handle the event queue.
+-   A reference to auxilliary data for the methods.
+-   An array of subscribers, which are simple callbacks taking a reference
+    to an event structure, along with auxilliary arguments.
+-   Buckets of events, one bucket for each priority level.  The number of
+    buckets depends on how many priority levels are implemented by the
+    functions of the method structure.
+    If there are zero priority levels, there are no buckets at all, and
+    events "added" to such a queue structure are always dispatched
+    immediately, if the function used to "add" the event supports that.
+
+For each event added to the queue, there is associated data:
+
+-   *[optional]* A time value, to indicate when the event is to be fired
+    off (i.e. when it's going to be passed to subscribers)
+
+    This may be zero to indicate that it should be fired off as soon as
+    possible, and will otherwise be compared to the current time to figure
+    out if it should be fired off or should remain in the event queue for a
+    bit longer.
+
+    The time value allows for time delay events, where the subscriber that
+    listens to them can simply generate a new time delay event, thereby
+    ensuring that they happen at regular intervals without any extra help
+    from outside calls.
+
+#### Associated structures and services
+
+The ctrl() function in the method structure of an event queue may manipulate
+the auxilliary data, as well as perform extraordinary actions that affect
+other structures, such as other event queues, as well as the event queue
+that such calls originate from.
+
+Similarly, subscribers may affect other structures, even other event queues,
+assuming that references to them are passed as part of the subscriber data.
+
+This can be useful for "services" such as a loss detector, congestion control,
+etc.
+
+A very concrete example:
+
+Let's assume that packet retransmit data is stored in a structure somewhere
+(lets call it a retransmission queue), to be retrasmitted at some point.
+
+Two things can happen:
+
+1.  An ACK is received, triggering a READ event with ACK payload.
+2.  A timer event for retransmission is fired.
+
+The action to perform for each case is:
+
+1.  The event subscriber finds the retransmit data corresponding to the ACK
+    data and removes it from the the retransmission queue.
+2.  The timer event subscriber retransmits packets according to every items
+    it finds in the retransmission queue.
```

Comment:
Surely not everything it finds in the retransmission queue? Only those things that have "expired"?